### PR TITLE
xds: Supports injecting custom ldsResourceNameResolvers

### LINKED ISSUE #12667: gRPC Java - Is it possible to bind listener to an real IP (not 0.0.0.0) - building xDS enabled gRPC Server
I am building an xDS enabled gRPC Server. The gRPC Server is sending "0.0.0.0:#port" to xDS during LDS. 

Trying to find, how to pass socket address Or is it possible to pass address/IP.

#### gRPC Java file - grpc-java/xds/src/main/java/io/grpc/xds/XdsServerBuilder.java
```

  /** Creates a gRPC server builder for the given port. */
  public static XdsServerBuilder forPort(int port, ServerCredentials serverCredentials) {
    checkNotNull(serverCredentials, "serverCredentials");
    InternalProtocolNegotiator.ServerFactory originalNegotiatorFactory =
            InternalNettyServerCredentials.toNegotiator(serverCredentials);
    ServerCredentials wrappedCredentials = InternalNettyServerCredentials.create(
            new FilterChainMatchingNegotiatorServerFactory(originalNegotiatorFactory));
    NettyServerBuilder nettyDelegate = NettyServerBuilder.forPort(port, wrappedCredentials);
    return new XdsServerBuilder(nettyDelegate, port);
  }

  @Override
  public Server build() {
    checkState(isServerBuilt.compareAndSet(false, true), "Server already built!");
    FilterChainSelectorManager filterChainSelectorManager = new FilterChainSelectorManager();
    Attributes.Builder builder = Attributes.newBuilder()
            .set(ATTR_FILTER_CHAIN_SELECTOR_MANAGER, filterChainSelectorManager);
    if (drainGraceTimeUnit != null) {
      builder.set(ATTR_DRAIN_GRACE_NANOS, drainGraceTimeUnit.toNanos(drainGraceTime));
    }
    InternalNettyServerBuilder.eagAttributes(delegate, builder.build());
    return new XdsServerWrapper("0.0.0.0:" + port, delegate, xdsServingStatusListener,
            filterChainSelectorManager, xdsClientPoolFactory, bootstrapOverride, filterRegistry);
  }

```

The IP in above code is hardcoded to "0.0.0.0". 

My gRPC server code - 
```
server = XdsServerBuilder.forPort("5001",  InsecureServerCredentials.create())
               .addService(new EchoServiceImp())
               .addService(healthSattusManager.getHealthService())
               .addService(ProtoReflectionService.newInstance())
               .build()
               .start();

```

I do not see "XdsServerBuilder.forAddress" function. 

If passing address is not possible, then how to pass the gRPC server IP (not "0.0.0.0") to xDS via ways defined in [A36](https://github.com/grpc/proposal/blob/master/A36-xds-for-servers.md)

Passing a machine ip when binding listener will enable populate address in LDS to xDS: 
"server_listener_resource_name_template": "example/resource/%s"

### PR DESCRIPTION
**Pull request by ran-su**
Also add support for creating XdsServerBuilder with SocketAddresses.

Fixes #12667

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsServerBuilder.java [hunk 1]
```diff
@@ -134,6 +139,24 @@ public static XdsServerBuilder forPort(int port, ServerCredentials serverCredent
     return new XdsServerBuilder(nettyDelegate, port);
   }
 
+  /** Creates a gRPC server builder for the given address. */
+  public static XdsServerBuilder forAddress(
```
Comment (kannanjgithub) [inline xds/src/main/java/io/grpc/xds/XdsServerBuilder.java:143]: Add unit tests for this new method, for testing both Inet and custom SocketAddress, and InetAddress with and without port number suffixed, and null arguments disallowed.

Comment (ran-su) [reply xds/src/main/java/io/grpc/xds/XdsServerBuilder.java:143]: Added tests for `forAddress()`.

File: xds/src/test/java/io/grpc/xds/XdsServerBuilderTest.java [hunk 2]
```diff
@@ -276,6 +277,50 @@ public void xdsServerStartSecondUpdateAndError()
     verifyServer(null, mockXdsServingStatusListener, null);
   }
 
+  @Test
```
Comment (kannanjgithub) [inline xds/src/test/java/io/grpc/xds/XdsServerBuilderTest.java:289]: 1. Resource Leakage (Missing Cleanup Registration)
   * The Problem: The forAddress() test builds three separate XdsServerWrapper instances using builder.build(). None of these instances are registered with the cleanupRule (e.g., cleanupRule.register(...)). 
   * The Impact: Even when not explicitly started, building the server constructs the delegate Netty server, which can allocate system resources, thread pools, or file descriptors. Without registering them to cleanupRule, these resources are leaked
 across tests.

  ---

  2. Fragile Reflection on Internal Fields
   * The tests use reflection to inspect the private field listenerAddress inside XdsServerWrapper:
```
java.lang.reflect.Field listenerAddressField =
XdsServerWrapper.class.getDeclaredField("listenerAddress");
      listenerAddressField.setAccessible(true);
```
   It is possible to entirely on public APIs, checking actual server behavior and configuration like xdsServer.getPort().

  ---

  3. Monolithic Anti-Pattern
   * It is better to isolate each scenario into its own descriptive, single-responsibility test method (e.g., forAddress_nullCredentials_throws, forAddress_inetSocketAddress_startsAndShutsDown).

  ---

  4. Superficial Inspection vs. Behavioral Verification
   * The Problem: forAddress() only checks the static, built state of the wrapper using reflection. It never verifies that the constructed server actually works.
   * The Impact: You could easily write a bug in XdsServerBuilder that populates listenerAddress correctly but fails to properly initialize netty delegate handlers or register the correct TLS/credential contexts. forAddress() would pass, but the
     server would fail at runtime.
   * Suggestion: We could do comprehensive behavioral validation by starting the server, pushing simulated XDS updates (e.g., generateListenerUpdate), and verifying that the server actively serves and shuts down cleanly.

Below is an illustration of what it might look like:
```
  private void buildServerForAddress(
      SocketAddress address,
      XdsServerBuilder.XdsServingStatusListener xdsServingStatusListener)
      throws IOException {
    builder =
        XdsServerBuilder.forAddress(
            address, XdsServerCredentials.create(InsecureServerCredentials.create()));
    builder.xdsClientPoolFactory(xdsClientPoolFactory);
    builder.overrideBootstrapForTest(XdsServerTestHelper.RAW_BOOTSTRAP);
    if (xdsServingStatusListener != null) {
      builder.xdsServingStatusListener(xdsServingStatusListener);
    }
    tlsContextManager = mock(TlsContextManager.class);
    xdsServer = cleanupRule.register((XdsServerWrapper) builder.build());
  }

  @Test
  public void forAddress_nullCredentials_throws() {
    SocketAddress address = new InetSocketAddress(0);
    try {
      XdsServerBuilder.forAddress(address, null);
      fail("should have thrown");
    } catch (NullPointerException e) {
      assertThat(e).hasMessageThat().contains("serverCredentials");
    }
  }

  @Test
  public void forAddress_nullAddress_throws() {
    try {
      XdsServerBuilder.forAddress(null, XdsServerCredentials.create(InsecureServerCredentials.create()));
      fail("should have thrown");
    } catch (NullPointerException e) {
      // expected
    }
  }

  @Test
  public void forAddress_inetSocketAddress_startsAndShutsDown()
      throws IOException, InterruptedException, TimeoutException, ExecutionException {
    InetSocketAddress address = new InetSocketAddress(0);
    buildServerForAddress(address, null);
    Future<Throwable> future = startServerAsync();
    XdsServerTestHelper.generateListenerUpdate(
        xdsClient,
        CommonTlsContextTestsUtil.buildTestInternalDownstreamTlsContext("CERT1", "VA1"),
        tlsContextManager);
    verifyServer(future, null, null);
    verifyShutdown();
  }

  @Test
  public void forAddress_inetSocketAddress_withSpecificPort()
      throws IOException, InterruptedException, TimeoutException, ExecutionException {
    ServerSocket serverSocket = new ServerSocket(0);
    int port = serverSocket.getLocalPort();
    serverSocket.close();

    InetSocketAddress address = new InetSocketAddress("0.0.0.0", port);
    buildServerForAddress(address, null);
    Future<Throwable> future = startServerAsync();
    XdsServerTestHelper.generateListenerUpdate(
        xdsClient,
        CommonTlsContextTestsUtil.buildTestInternalDownstreamTlsContext("CERT1", "VA1"),
        tlsContextManager);
    verifyServer(future, null, null);
    assertThat(xdsServer.getPort()).isEqualTo(port);
    verifyShutdown();
  }

  @Test
  public void forAddress_customSocketAddress_builds() throws IOException {
    SocketAddress address = new SocketAddress() {};
    builder =
        XdsServerBuilder.forAddress(
            address, XdsServerCredentials.create(InsecureServerCredentials.create()));
    builder.xdsClientPoolFactory(xdsClientPoolFactory);
    builder.overrideBootstrapForTest(XdsServerTestHelper.RAW_BOOTSTRAP);
    xdsServer = cleanupRule.register((XdsServerWrapper) builder.build());
    assertThat(xdsServer).isNotNull();
  }
```

**Review by kannanjgithub — APPROVED** [review]
